File: video-editor/scripts/lib/platform.js
/* Cross-platform helpers (Windows · Linux) shared by the Node scripts. */
'use strict';
const fs = require('fs');
const path = require('path');
const { execFileSync } = require('child_process');

/* The skill root — same arithmetic as VEVO_SKILL_DIR in platform.sh:
   scripts/lib/platform.js -> scripts/lib -> scripts -> root. */
function skillDir() {
  return path.join(__dirname, '..', '..');
}

/* ffmpeg / ffprobe — honour $VEVO_FFMPEG / $VEVO_FFPROBE (same names platform.sh exports),
   else the bare name (PATH resolves it). Issue #44. */
function ffmpegPath()  { return process.env.VEVO_FFMPEG  || 'ffmpeg'; }
function ffprobePath() { return process.env.VEVO_FFPROBE || 'ffprobe'; }

function commandExists(cmd) {
  try {
    execFileSync(process.platform === 'win32' ? 'where' : 'which', [cmd], { stdio: 'ignore' });
    return true;
  } catch (_) { return false; }
}

/* Isolated Python interpreter — same logic as VEVO_PY in platform.sh:
   uv (skill project) > skill .venv > system python3 as a last resort.
   Returns [cmd, ...leadingArgs] to prefix to the real arguments:
     const c = pythonCmd(); execFileSync(c[0], [...c.slice(1), 'script.py', arg]) */
function pythonCmd() {
  const skill = skillDir();
  if (commandExists('uv')) return ['uv', 'run', '--project', skill, 'python'];
  const venvPy = process.platform === 'win32'
    ? path.join(skill, '.venv', 'Scripts', 'python.exe')
    : path.join(skill, '.venv', 'bin', 'python');
  if (fs.existsSync(venvPy)) return [venvPy];
  return ['python3'];
}

module.exports = { pythonCmd, skillDir, ffmpegPath, ffprobePath };

File: video-editor/scripts/lib/platform.py
# -*- coding: utf-8 -*-
"""Cross-platform helpers shared by the Python scripts — the mirror of
`lib/platform.sh` (`VEVO_*`) and `lib/platform.js` for the Python side.

Only the ffmpeg / ffprobe resolver for now (issue #44): the ~16 Python call sites used
the bare string `"ffmpeg"` / `"ffprobe"`, so there was no single place to point them at a
specific binary. `FFMPEG` / `FFPROBE` honour `$VEVO_FFMPEG` / `$VEVO_FFPROBE` (same names
`platform.sh` exports), then fall back to the name on PATH.

    from lib import platform as _plat
    subprocess.run([_plat.FFMPEG, "-v", "error", "-i", src, ...])
"""
import os
import shutil


def _tool(name, env):
    """Absolute path to `name`, honouring $<env>; the bare name if PATH can't resolve it
    yet (it may be installed by the time the subprocess runs)."""
    want = os.environ.get(env) or name
    return shutil.which(want) or want


def ffmpeg():
    return _tool("ffmpeg", "VEVO_FFMPEG")


def ffprobe():
    return _tool("ffprobe", "VEVO_FFPROBE")


FFMPEG = ffmpeg()
FFPROBE = ffprobe()
